File: Assignment6_2/src/main/java/view/CurrencyConverterView.java
package view;

import controller.CurrencyConverterController;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;

import java.util.List;

public class CurrencyConverterView extends Application {

    private CurrencyConverterController controller;

    private ChoiceBox choiceBoxSource = new ChoiceBox<>();
    private ChoiceBox choiceBoxTarget = new ChoiceBox<>();
    private TextField source = new TextField();
    private TextField target = new TextField();

    Alert alert = new Alert(Alert.AlertType.ERROR, "Please check the number format or select a currency.");

    @Override
    public void start(Stage stage) {
        stage.setTitle("Currency Converter");
        stage.setResizable(false);

        HBox hBox = new HBox();
        hBox.setAlignment(Pos.BOTTOM_CENTER);

        source.setMinWidth(200);
        hBox.getChildren().add(source);
        HBox.setMargin(source, new Insets(10, 10, 10, 10));

        VBox vBox1 = new VBox();
        vBox1.getChildren().add(new Label("Select Currency"));
        vBox1.getChildren().add(choiceBoxSource);
        vBox1.setAlignment(Pos.CENTER);
        VBox.setMargin(choiceBoxSource, new Insets(2, 0, 0, 0));

        hBox.getChildren().add(vBox1);
        HBox.setMargin(vBox1, new Insets(5, 10, 10, 0));

        Button convertButton = new Button("Convert");
        convertButton.getStyleClass().add("convert-button");
        hBox.getChildren().add(convertButton);
        HBox.setMargin(convertButton, new Insets(10, 10, 10, 0));

        choiceBoxSource.setOnAction(event -> {
            controller.onSourceChoiceBoxChange();
            choiceBoxSource.getItems().remove("Select");
        });

        target.setMinWidth(200);
        hBox.getChildren().add(target);
        HBox.setMargin(target, new Insets(10, 10, 10, 0));

        VBox vBox2 = new VBox();
        vBox2.getChildren().add(new Label("Select Currency"));
        vBox2.getChildren().add(choiceBoxTarget);
        vBox2.setAlignment(Pos.CENTER);
        VBox.setMargin(choiceBoxTarget, new Insets(2, 0, 0, 0));
        hBox.getChildren().add(vBox2);
        HBox.setMargin(vBox2, new Insets(5, 10, 10, 0));

        source.setOnKeyTyped(event -> {
            target.setText("");
        });

        choiceBoxTarget.setOnAction(event -> {
            controller.onTargetChoiceBoxChange();
            choiceBoxTarget.getItems().remove("Select");
        });

        convertButton.setOnAction(event -> {
            controller.convertButtonPressed();
        });

        hBox.getStyleClass().add("hBox");

        Scene scene = new Scene(hBox);

        scene.getStylesheets().add(getClass().getResource("/style.css").toExternalForm());

        stage.setScene(scene);

        target.setEditable(false);
        stage.show();
    }

    @Override
    public void init() {
        controller = new CurrencyConverterController(this);
    }

    public void setChoiceBoxSource(List<String> list) {
        choiceBoxSource.setItems(FXCollections.observableList(list));
    }

    public void setChoiceBoxTarget(List<String> list) {
        choiceBoxTarget.setItems(FXCollections.observableList(list));
    }

    public void setChoiceBoxSourceValue(String s) {
        choiceBoxSource.setValue(s);
    }

    public void setChoiceBoxTargetValue(String s) {
        choiceBoxTarget.setValue(s);
    }

    public String getSource() {
        return source.getText();
    }

    public String getTarget() {
        return target.getText();
    }

    public void setSource(String s) {
        source.setText(s);
    }

    public void setSourcePlaceholder(String s) {
        source.setPromptText(s);
    }

    public void setTarget(String s) {
        target.setText(s);
    }

    public void setTargetPlaceholder(String s) {
        target.setPromptText(s);
    }

    public void showAlert() {
        alert.show();
    }

    public String[] getChoiceBoxes() {
        return new String[]{(String) choiceBoxSource.getValue(), (String) choiceBoxTarget.getValue()};
    }
}

File: Assignment7_2_Kotlin/src/main/kotlin/application/view/CurrencyConverterView.kt
package application.view

import javafx.application.Application
import javafx.application.Platform
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.event.ActionEvent
import javafx.event.Event
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.input.KeyEvent
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.Stage
import java.util.function.Consumer


class CurrencyConverterView : Application() {
    private var controller: application.controller.CurrencyConverterController? = null

    private val choiceBoxSource: ChoiceBox<String> = ChoiceBox<String>()
    private val choiceBoxTarget: ChoiceBox<String> = ChoiceBox<String>()
    private val source = TextField()
    private val target = TextField()

    private val alert = Alert(Alert.AlertType.ERROR, "Please select currency or fill the number field.")

    private val SQLAlert = Alert(Alert.AlertType.ERROR)

    private var stage: Stage? = null

    override fun start(primaryStage: Stage?) {
        stage = Stage()
        stage!!.setTitle("Currency Converter")
        stage!!.setResizable(false)
        target.setEditable(false)


        val hBox = HBox()
        hBox.setAlignment(Pos.BOTTOM_CENTER)

        source.setMinWidth(200.0)
        hBox.children.add(source)
        HBox.setMargin(source, Insets(10.0, 10.0, 10.0, 10.0))

        val vBox1 = VBox()
        vBox1.children.add(Label("Select Currency"))
        vBox1.children.add(choiceBoxSource)
        vBox1.setAlignment(Pos.CENTER)
        VBox.setMargin(choiceBoxSource, Insets(2.0, 0.0, 0.0, 0.0))

        hBox.children.add(vBox1)
        HBox.setMargin(vBox1, Insets(5.0, 10.0, 10.0, 0.0))

        val convertButton = Button("Convert")
        convertButton.styleClass.add("convert-button")
        hBox.children.add(convertButton)
        HBox.setMargin(convertButton, Insets(10.0, 10.0, 10.0, 0.0))

        choiceBoxSource.setOnAction(EventHandler { event: Event? ->
            controller!!.onSourceChoiceBoxChange()
            choiceBoxSource.getItems().remove("Select")
        })

        target.minWidth = 200.0
        hBox.children.add(target)
        HBox.setMargin(target, Insets(10.0, 10.0, 10.0, 0.0))

        val vBox2 = VBox()
        vBox2.children.add(Label("Select Currency"))
        vBox2.children.add(choiceBoxTarget)
        vBox2.setAlignment(Pos.CENTER)
        VBox.setMargin(choiceBoxTarget, Insets(2.0, 0.0, 0.0, 0.0))
        hBox.children.add(vBox2)
        HBox.setMargin(vBox2, Insets(5.0, 10.0, 10.0, 0.0))

        source.setOnKeyTyped(EventHandler { event: KeyEvent? ->
            target.setText("")
        })

        // input filtering
        source.textProperty()
            .addListener(ChangeListener { observable: ObservableValue<out String?>?, oldValue: String?, newValue: String? ->
                // NOTE: ChatGPT generated regex
                if (!newValue!!.matches("\\d*([.,]\\d{0,2})?".toRegex())) {
                    source.text = oldValue
                }
            })

        choiceBoxTarget.onAction = EventHandler { event: Event? ->
            controller!!.onTargetChoiceBoxChange()
            choiceBoxTarget.getItems().remove("Select")
        }

        convertButton.onAction = EventHandler { event: ActionEvent? ->
            controller!!.convertButtonPressed()
        }

        hBox.styleClass.add("hBox")

        val scene = Scene(hBox)

        scene.stylesheets.add(javaClass.getResource("/style.css")?.toExternalForm())

        stage!!.setScene(scene)

        if (!SQLAlert.isShowing) {
            stage!!.show()
        }
    }

    override fun init() {
        controller = application.controller.CurrencyConverterController(this)
    }

    fun setChoiceBoxSource(list: MutableList<String?>) {
        choiceBoxSource.items = FXCollections.observableList(list)
    }

    fun setChoiceBoxTarget(list: MutableList<String?>) {
        choiceBoxTarget.items = FXCollections.observableList(list)
    }

    fun setChoiceBoxSourceValue(s: String?) {
        choiceBoxSource.value = s
    }

    fun setChoiceBoxTargetValue(s: String?) {
        choiceBoxTarget.value = s
    }

    fun getSource(): String? {
        return source.text
    }

    fun getTarget(): String? {
        return target.text
    }

    fun setSource(s: String?) {
        source.setText(s)
    }

    fun setSourcePlaceholder(s: String?) {
        source.setPromptText(s)
    }

    fun setTarget(s: String?) {
        target.setText(s)
    }

    fun setTargetPlaceholder(s: String?) {
        target.setPromptText(s)
    }

    fun showAlert() {
        Platform.runLater(Runnable {
            alert.show()
        })
    }

    val choiceBoxes: Array<String?>
        get() = arrayOf<String?>(choiceBoxSource.getValue() as String?, choiceBoxTarget.getValue() as String?)

    fun showSQLError() {
        println("AHHHH")
        Platform.runLater(Runnable {
            SQLAlert.setTitle("Database Error")
            SQLAlert.setHeaderText("Cannot retrieve currency data")
            SQLAlert.setContentText("Database not found or no currency data")
            SQLAlert.showAndWait().ifPresent { buttonType: ButtonType? ->
                Platform.exit()
            }
        })
    }
}